RFC.md

```diff
@@ -0,0 +1,466 @@
+## Context
+The majority of SOTA multimodal models coming out these days are not trained end to end as uniform models, but instead a separate model is trained for each modality and then these models are fused together. In the simplest case, you have an LLM trained on text and an image encoder such as CLIP. You then need to combine these pre-trained models and finetune them so that the LLM can understand the embeddings from the image encoder.
+
+There are many different ways to combine them, and a detailed breakdown of these approaches can be found [here](https://arxiv.org/pdf/2405.17927). At a high level, there are two overarching architectures that most models can be grouped into. **EarlyFusion** and **DeepFusion**. **EarlyFusion** defines a model where a mapping is learned from encoder outputs to token embeddings. Then encoded inputs are converted to tokens and fed into the LLM the same as text tokens. On the other hand, in **DeepFusion**, the output of the encoder is injected directly into the intermediate layers of the LLM, typically with cross attention, following the same approach for encoder decoder language models.
+
+To work with these models, there are several considerations. If you have two+ pre-trained models, you may want to be able to fuse them and add some additional learnable parameters (we will call **fusion parameters**). For this you want to be able to define the fused model but be able to load in the individual checkpoints. You also want to be able to determine which parameters/sub-modules to make trainable and which to keep frozen. Apart from directly fusing models, there are a lot of open models (like Llava, PaliGemma, Idefics, Phi Vision, etc) that have already been fused, but you might want to further finetune. In this case you want to be able to load in the full fused checkpoint and further finetuning on the models, either on all the weights or using PeFT techniques. To handle all of this, and keep uniform recipes/utils, it would be ideal if the fused models could behave as a single model but then easily decompose into their individual parts.
+
+**TLDR**
+- Multimodal models are primarily DeepFusion and EarlyFusion architectures
+- Modules/utils for fusing and un-fusing MM models can allow for fine-tuning a variety of MM architectures in a uniform way.
+
+## Proposal
+
+**High Level List:**
+- Introduce a modules/model_fusion folder for handling fused models
+- fusion_models introduces two top level wrapper modules that combine a single encoder and decoder in two different ways. This allows the two models to be used as one model.
+- fusion_layer and fusion_embed are fusion modules (similar to PeFT modules) that introduce extra trainable parameters to the pre-trained fused models
+- fusion_utils is based on peft_utils and provides functions for accessing and modifying fusion parameters
+
+**Folder Structure for new modules/functions**
+```
+torchtune/
+	modules/
+		model_fusion/
```

Comment:
I understand that this is the "technical" name, but I feel like most people know these utilities as parts of multimodal_llms. I would be in favor of naming the folder accordingly.

Reply:
The reason I didn't use multimodal anywhere is because none of this technically pertains to multimodal but instead it's how how to handle combining pre-trained models. I could be convinced to replace **fusion** everywhere with **multimodal** but I'm going to be upset when I have to use these modules for fusing two models together that have nothing to do with multimodality 😆